demos/vlc-dag/src/lib.rs

```diff
@@ -0,0 +1,720 @@
+    }
+
+    /// Terminate a running accumulator server.
+    pub async fn terminate(&mut self, index: usize) {
+        let msg = Message::Terminate;
+        self.socket
+            .send_to(
+                serde_json::to_string(&msg).unwrap().as_bytes(),
+                self.config.server_addrs[index],
+            )
+            .await
+            .unwrap();
+    }
+}
+
+/// State of a server node.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ServerState {
+    clock_info: ClockInfo,
+    id: u128,
+    items: Vec<String>, // BtreeSet or HashSet can't keep event inserted order.
+    clock_to_eventid: HashMap<String, String>, // clock_id to event_id
+    message_id_set: BTreeSet<String>,
+}
+
+impl ServerState {
+    /// Create a new server state.
+    fn new(id: u128) -> Self {
+
+        Self {
+            clock_info: ClockInfo {
+                clock: Clock::new(),
+                id,
+                message_id: "".to_string(),
+                count: 0,
+                create_at: get_time_ms(),
+            },
+            id,
+            items: Vec::new(),
+            clock_to_eventid: HashMap::new(),
+            message_id_set: BTreeSet::new(),
+        }
+    }
+
+    /// Add items into the state. Returns true if resulting in a new state.
+    fn add(&mut self, items: Vec<String>) -> bool {
+        if items.len() < 1 {
+            false
+        } else {
+            for it in items.clone() {
+                // filter replicate message id
+                if self.message_id_set.contains(&it) {
+                    continue;
+                }
+                self.items.push(it.clone());
+                self.message_id_set.insert(it);
+            }
+            self.clock_info.clock.inc(self.id);
+            self.clock_info.count += 1;
+            if let Some(last) = items.last() {
+                self.clock_info.message_id = last.to_string();
+            }
+            true
+        }
+    }
+
+    /// Deprecated
+    /// Merge another ServerState into the current state. Returns true if
+    /// resulting in a new state (different from current and received
+    /// state). return first bool is need_broadcast, second bool is need_log
+    fn merge(&mut self, other: &Self) -> (bool, bool) {
+        match self.clock_info.clock.partial_cmp(&other.clock_info.clock) {
+            Some(cmp::Ordering::Equal) => return (false, false),
+            Some(cmp::Ordering::Greater) => return (false, false),
+            Some(cmp::Ordering::Less) => {
+                self.clock_info = other.clock_info.clone();
+                self.items = other.items.clone();
+                (false, true)
+            }
+            None => {
+                self.clock_info.clock.merge(&vec![&other.clock_info.clock]);
+                let add = self.add(other.items.clone());
+                (add, add)
+            }
+        }
+    }
+
+    fn handle_event_trigger(&mut self, msg: EventTrigger) -> Option<ServerMessage> {
```

Comment:
```suggestion
    fn handle_event_trigger(&mut self, msg: EventTrigger) -> Option<ServerMessage> {
    match self.clock_info.clock.partial_cmp(&msg.clock_info.clock) {
        Some(cmp::Ordering::Less) => {
            let self_clock_info = ClockInfo::new(self.clock_info.clock.clone(), self.id, self.clock_info.message_id.clone(), 0);
            let req = ServerMessage::DiffReq(DiffReq { from_clock: self_clock_info, to:  msg.clock_info.id });
            Some(req)
        }
        None => {
            let base = self.clock_info.clock.base_common(&msg.clock_info.clock);
            if base.is_genesis() || self.clock_to_eventid.get(&base.index_key()).map_or(false, |start_msg_id| get_suffix(&self.items, start_msg_id.to_string()).is_some()) {
                let req = ServerMessage::ActiveSync(ActiveSync { diffs: self.items.clone(), latest: self.clock_info.clone(), to: msg.clock_info.id });
                Some(req)
            } else {
                None
            }
        }
        _ => None,
    }
}
```

Reply:
Expose all of the conditions will be more information, explicitly.